### ISSUE TITLE
dotnet run leaves Linux terminal in broken state after unhandled exception following Console.ReadLine

### ISSUE THREAD

**Issue by Encoded404** [dotnet/runtime#110502]
### Description

When running a console application with dotnet run on Linux, if an exception occurs after a Console.ReadLine call, the terminal is left in a broken state. Specifically, the terminal no longer echoes user input, and the issue persists until the terminal is reset using the reset command.

### Reproduction Steps

1. Create a new console app using `dotnet new` console.
2. Replace the contents of Program.cs with the following code:
```
Console.ReadLine();

// Intentional crash
int[] arr = new int[1];
arr[100] = 10;
```

3. Run the application using `dotnet run`.
4. type a letter and press enter
5. Observe that after the program crashes, the terminal no longer echoes user input.

### Expected behavior

The terminal should continue to function normally, including echoing user input, even after the program exits due to an unhandled exception.

### Actual behavior

The terminal stops echoing user input after the program crashes, requiring the user to run the reset command or reopen the terminal to restore normal functionality.

### Known Workarounds

1. Use dotnet build to compile the program and then execute the binary directly (./bin/Debug/net8.0/YourProjectName).
2. After the terminal enters the broken state, use the reset command to restore normal functionality.

### Configuration

- dotnet version: 8.0.111
- OS: Fedora Linux 41 (version 6.11.10-300.fc41.x86_64)
- architecture: x86_64 / x64
- Additional Info: i haven't tested on other Linux distros

**Comment by Encoded404** [comment]
After further testing, I discovered that the issue also occurs when running the compiled binary directly through a bash script, such as:

```
#!/bin/bash
dotnet build
./bin/Debug/net8.0/ProjectName
```

When the script is executed, the terminal is still left in a broken state (input is not echoed). However, running the same commands manually in the terminal (dotnet build followed by ./bin/Debug/net8.0/ProjectName) does not cause this behavior.

you could add a reset or stty sane to the end of the bash script to reset the console automatically.

This does, however, suggest that the issue might be specific to my machine or environment. To confirm this, I plan to test it on another Linux distribution on a different computer I have.

**Comment by Encoded404** [comment]
I have confirmed that this issue is not limited to my local machine. It also occurs on a headless Ubuntu server over ssh with the following configuration:
dotnet Version: 8.0.110
Ubuntu Version: Ubuntu 24.04.1 LTS

The issue occurs under the same conditions when using either dotnet run or running the compiled binary via a bash script. The terminal is left in a broken state where input is not echoed until reset or stty sane is run.

**Comment by arctixdev** [comment]
I am experiencing the same, also on Fedora 41 with dotnet 8.0.111, but only using bash, I can't reproduce using ZSH as the shell.

**Comment by arctixdev** [comment]
After further investigation, it seems that the TTY settings are not restored to their default state when a .NET program crashes.

To show this, I captured the TTY settings before and after running the simple .NET program that crashes due to an unhandled exception:

```bash
$ stty -a > before.log
$ dotnet run
test
test
Unhandled exception. System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at Program.<Main>$(String[] args) in /home/jonathan/tmp/dotnettest/Program.cs:line 5
$ stty -a > after.log
$ diff before.log after.log
6c6
< -ignbrk brkint ignpar -parmrk -inpck -istrip -inlcr -igncr icrnl -ixon -ixoff
---
> -ignbrk brkint ignpar -parmrk -inpck -istrip -inlcr -igncr -icrnl -ixon -ixoff
9c9
< isig icanon iexten echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
---
> isig -icanon -iexten -echo echoe echok -echonl -noflsh -xcase -tostop -echoprt
```

As seen in the output, the following TTY settings are disabled after the crash:
- `icrnl`
- `icanon`
- `iexten`
- `echo`

This change leaves the terminal in a non-standard state. Running `stty sane` after the crash restores these settings, but this manual step shouldn't be necessary.

**Expected Behavior**:  
The .NET runtime should restore the TTY to its default state automatically, as it does when the program exits cleanly. 
**Proposed Fix**:  
The .NET runtime should ensure TTY cleanup logic is executed consistently, even during unhandled exceptions or abrupt termination.

**Comment by tmds** [comment]
Yes, we only restore the terminal on a clean exit currently by registering an `atexit` handler:

https://github.com/dotnet/runtime/blob/4b3902a13b094e1c090b38861ecc3bd793ec46d0/src/native/libs/System.Native/pal_console.c#L451

> The .NET runtime should ensure TTY cleanup logic is executed consistently, even during unhandled exceptions or abrupt termination.

If we call `UninitializeTerminal` for signals for abnormal exit, we should get this behavior.

The trickiness is mostly around understanding the interaction with the runtime which also cares and (partly) handles these signals.

cc @adamsitnik @stephentoub @jkotas

**Comment by tmds** [comment]
PS: having a try-catch in your `Main` method may be a simple way to workaround the issue.
```cs
try
{
  ...
  return 0;
}
catch (Exception ex)
{
  Console.Error.WriteLine("Unhandled exception:");
  Console.Error.WriteLine(ex);
  return 1;
}
```

**Comment by adamsitnik** [comment]
> The trickiness is mostly around understanding the interaction with the runtime which also cares and (partly) handles these signals.

I don't have a good understanding of that interaction (as of now), but if we want to give it a try, now it's a good time for this, as we have shipped .NET 9 and there is plenty of time to test how given fix would affect .NET apps starting from .NET 10 preview 1.

**Comment by jkotas** [comment]
> The trickiness is mostly around understanding the interaction with the runtime which also cares and (partly) handles these signals.

Unhandled exceptions call `abort`. Are you thinking about adding the cleanup as a SIGABRT handler?

Alternatively, the cleanup for unhandled exceptions can be done in C# by registering AppDomain.UnhandledException event handler.

I would limit the fix to unhandled managed exception. There are always going to be fatal crashes that leave the terminal in bad state, it is impossible to make this perfect.

**Comment by tmds** [comment]
> Alternatively, the cleanup for unhandled exceptions can be done in C# by registering AppDomain.UnhandledException event handler.

👍 I like this option. It is much simpler than dealing with signal handlers.

> I would limit the fix to unhandled managed exception. There are always going to be fatal crashes that leave the terminal in bad state, it is impossible to make this perfect.

Agreed.

@adamsitnik, @jkotas I can look into implementing this.

**Comment by adamsitnik** [comment]
> I can look into implementing this.

Great, thanks a lot! BTW I will be able to provide code reviews whole next week, then I'll be gone for Christmas break

**Comment by evankyle** [comment]
Another workaround, add echo -ne "\033[?1l" to your prompt command.
